packages/big-design/src/components/Worksheet/utils.ts

```diff
@@ -18,8 +18,32 @@ export const mergeCells = <T extends unknown>(oldCells: Cell<T>[], newCells: Cel
     [...oldCells], // Note: returns a new array every time
   );
 
-export const deleteCells = <T extends unknown>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
+export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
   oldCells.filter(
-    (oldCell) =>
-      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
+    (oldCell) => !newCells.find((newCell) => newCell.hash === oldCell.hash && newCell.rowIndex === oldCell.rowIndex),
   );
+
+export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
+  editedCells.reduce<T[]>((accum, editedCell) => {
+    const row = rows[editedCell.rowIndex];
+
+    if (accum.find((editedRow) => editedRow === row)) {
+      return accum;
+    } else {
+      return [...accum, row];
+    }
+  }, []);
+
+export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) =>
+  invalidCells.reduce<WorksheetError<T>[]>((accum, invalidCell) => {
+    const row = rows[invalidCell.rowIndex];
+    const rowIndex = accum.findIndex((invalidRow) => invalidRow.item === row);
+
+    if (rowIndex >= 0) {
+      accum[rowIndex].errors = [...accum[rowIndex].errors, invalidCell.hash];
+
+      return accum;
+    } else {
+      return [...accum, { item: row, errors: [invalidCell['hash']] }];
+    }
+  }, []);
```

Comment:
I don't see how can simply `map` over the items since we could have 5 modified cells (in the same row) and return a single item. Unless you mean map and filter in conjunction and it gets more complicated than a reduce?